libsolidity/analysis/SyntaxChecker.cpp

```
@@ -210,36 +210,32 @@ bool SyntaxChecker::visit(PlaceholderStatement const&)
 	return true;
 }
 
+bool SyntaxChecker::visit(ContractDefinition const& _contract)
+{
+	ASTString const& contractName = _contract.name();
+	for (const FunctionDefinition* const function : _contract.definedFunctions())
+		if (function->name() == contractName)
+			m_errorReporter.syntaxError(function->location(),
+			"Defining constructors as functions with the same name as the contract is not allowed. "
+			"Use \"constructor(...) { ... }\" instead.");
+	return true;
+}
+
 bool SyntaxChecker::visit(FunctionDefinition const& _function)
 {
 	bool const v050 = m_sourceUnit->annotation().experimentalFeatures.count(ExperimentalFeature::V050);
 
 	if (v050 && _function.noVisibilitySpecified())
 		m_errorReporter.syntaxError(_function.location(), "No visibility specified.");
 
-	if (_function.isOldStyleConstructor())
-	{
-		if (v050)
-			m_errorReporter.syntaxError(
-				_function.location(),
-				"Functions are not allowed to have the same name as the contract. "
-				"If you intend this to be a constructor, use \"constructor(...) { ... }\" to define it."
-			);
-		else
-			m_errorReporter.warning(
-				_function.location(),
-				"Defining constructors as functions with the same name as the contract is deprecated. "
-				"Use \"constructor(...) { ... }\" instead."
-			);
-	}
 	if (!_function.isImplemented() && !_function.modifiers().empty())
 	{
 		if (v050)
 			m_errorReporter.syntaxError(_function.location(), "Functions without implementation cannot have modifiers.");
 		else
 			m_errorReporter.warning(_function.location(), "Modifiers of functions without implementation are ignored." );
 	}
-	if (_function.name() == "constructor")
+	if (_function.name() == "constructor" && !_function.isConstructor())
```

Please don't add `!_function.isConstructor()`. Currently, it looks like this is a good thing to add, but it might hide errors that are introduced when we change the logic of `isConstructor`.